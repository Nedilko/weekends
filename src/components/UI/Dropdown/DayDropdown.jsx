import PropTypes from 'prop-types'
import { getDay, DAYS } from '@utils/convertTime'
import Dropdown from '../Dropdown/Dropdown'

DayDropdown.propTypes = {
  selectedValue: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
}

function DayDropdown({ selectedValue, onChange }) {
  return (
    <Dropdown
      value={DAYS[selectedValue - 1]}
      items={DAYS}
      label="Day"
      onChange={(value) => onChange(getDay(value))}
    />
  )
}
export default DayDropdown
