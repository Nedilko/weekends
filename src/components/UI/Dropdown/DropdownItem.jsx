import PropTypes from 'prop-types'
import { useEffect, useRef } from 'react'

DropdownItem.propTypes = {
  handleSelect: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  isSelected: PropTypes.bool.isRequired,
}

function DropdownItem({ handleSelect, value, isSelected }) {
  const ref = useRef(null)

  useEffect(() => {
    if (isSelected && ref.current) {
      ref.current.scrollIntoView()
    }
  }, [isSelected])

  return (
    <option
      ref={ref}
      onClick={(e) => handleSelect(e.target.value)}
      value={value}
      data-testid="dropdown-item"
      className={`px-2 py-1 text-sm hover:bg-slate-200 dark:hover:bg-slate-700 ${
        isSelected ? 'font-regular' : 'font-thin'
      }`}
    >
      {value}
    </option>
  )
}

export default DropdownItem
