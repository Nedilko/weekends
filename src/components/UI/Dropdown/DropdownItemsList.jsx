import GeneralPanel from '@components/UI/GeneralPanel'
import PropTypes from 'prop-types'

DropdownItemsList.propTypes = {
  selectedValue: PropTypes.string,
  className: PropTypes.string,
  items: PropTypes.array,
  handleSelect: PropTypes.func,
}

function DropdownItemsList({ items, selectedValue, handleSelect, className }) {
  return (
    <GeneralPanel className={className}>
      {items.map((item, index) => (
        <option
          onClick={(e) => handleSelect(e.target.value)}
          key={index}
          value={item}
          className={`px-2 py-1 text-sm hover:bg-slate-200 dark:hover:bg-slate-700 ${
            selectedValue === item ? 'font-regular' : 'font-thin'
          }`}
        >
          {item}
        </option>
      ))}
    </GeneralPanel>
  )
}

export default DropdownItemsList
