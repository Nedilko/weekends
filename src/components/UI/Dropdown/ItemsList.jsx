import GeneralPanel from '@components/UI/GeneralPanel'
import PropTypes from 'prop-types'

ItemsList.propTypes = {
  selectedValue: PropTypes.string,
  items: PropTypes.array,
  handleSelect: PropTypes.func,
}

function ItemsList({ items, selectedValue, handleSelect }) {
  return (
    <GeneralPanel className="mt-1 max-h-28 w-28 cursor-pointer border border-gray-200">
      {items.map((item, index) => (
        <option
          onClick={(e) => handleSelect(e.target.value)}
          key={index}
          value={item}
          className={`px-2 py-1 text-sm hover:bg-slate-200 ${
            selectedValue === item ? 'font-regular' : 'font-thin'
          }`}
        >
          {item}
        </option>
      ))}
    </GeneralPanel>
  )
}

export default ItemsList
