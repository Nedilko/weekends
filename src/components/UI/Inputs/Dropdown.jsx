import { useEffect, useState, useRef } from 'react'

import PropTypes from 'prop-types'

Dropdown.propTypes = {
  value: PropTypes.string,
  items: PropTypes.array,
  label: PropTypes.string,
  onChange: PropTypes.func,
}

function Dropdown({ label, value, items, onChange }) {
  const dropdownMenu = useRef(null)
  const dropdownButton = useRef(null)

  const openClosedIcon = (isOpen) => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      className={`mr-1 h-4 w-4 text-gray-500 ${isOpen ? 'rotate-180' : ''}`}
      fill="none"
      viewBox="0 0 24 24"
      stroke="currentColor"
      strokeWidth="2"
    >
      <path strokeLinecap="round" strokeLinejoin="round" d="M19 9l-7 7-7-7" />
    </svg>
  )

  const [isActive, setIsActive] = useState(false)
  const [selectedValue, setSelectedValue] = useState(value)

  const handleClick = () => setIsActive(!isActive)

  const handleSelectItem = (e) => {
    setSelectedValue(e.target.value)
    setIsActive(false)
  }

  useEffect(() => {
    onChange(selectedValue)
  }, [selectedValue])

  useEffect(() => {
    function handleClickOutside(event) {
      if (
        dropdownMenu.current &&
        !dropdownMenu.current.contains(event.target) &&
        dropdownButton.current &&
        !dropdownButton.current.contains(event.target)
      ) {
        setIsActive(false)
      }
    }
    document.addEventListener('mousedown', handleClickOutside, true)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside, true)
    }
  }, [dropdownMenu])

  const drowpMenu = items.map((item, index) => (
    <option
      onClick={handleSelectItem}
      key={index}
      value={item}
      className={`px-2 py-1 text-sm hover:bg-slate-200 ${
        selectedValue === item ? 'font-regular' : 'font-thin'
      }`}
    >
      {item}
    </option>
  ))

  return (
    <div className="flex flex-row items-center">
      <div className="mr-2 flex items-center justify-center">{label}</div>
      <div className="block">
        <button
          ref={dropdownButton}
          type="button"
          className="flex h-6 w-28 items-center justify-between rounded-md border border-slate-100 pl-2 text-base font-thin focus:border-slate-300 focus:outline-none"
          onClick={handleClick}
        >
          {selectedValue}
          {openClosedIcon(isActive)}
        </button>
        {isActive && (
          <div
            ref={dropdownMenu}
            className="absolute mt-1 max-h-28 w-28 cursor-pointer overflow-auto rounded-md border border-gray-200 bg-white shadow-md"
          >
            {drowpMenu}
          </div>
        )}
      </div>
    </div>
  )
}

export default Dropdown
