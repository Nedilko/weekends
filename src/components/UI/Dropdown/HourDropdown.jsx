import PropTypes from 'prop-types'
import { getHour, HOURS } from '@utils/convertTime'
import Dropdown from '@UI/Dropdown/Dropdown'

HourDropdown.propTypes = {
  selectedValue: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
}

function HourDropdown({ selectedValue, onChange }) {
  return (
    <Dropdown
      value={HOURS[+selectedValue]}
      items={HOURS}
      label="Hour"
      onChange={(value) => onChange(getHour(value))}
    />
  )
}
export default HourDropdown
