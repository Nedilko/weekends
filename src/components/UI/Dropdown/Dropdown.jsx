import { useEffect, useState, useRef } from 'react'
import PropTypes from 'prop-types'
import DropdownItemsList from './DropdownItemsList'

Dropdown.propTypes = {
  value: PropTypes.string,
  items: PropTypes.array,
  label: PropTypes.string,
  onChange: PropTypes.func,
}

function Dropdown({ label, items, value, onChange }) {
  const dropdownMenu = useRef(null)

  const openClosedIcon = (isOpen) => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      className={`mr-1 h-4 w-4 text-gray-500 ${isOpen ? 'rotate-180' : ''}`}
      fill="none"
      viewBox="0 0 24 24"
      stroke="currentColor"
      strokeWidth="2"
    >
      <path strokeLinecap="round" strokeLinejoin="round" d="M19 9l-7 7-7-7" />
    </svg>
  )

  const [isOpen, setIsOpen] = useState(false)
  const [selectedValue, setSelectedValue] = useState(value)

  const handleSelectItem = (value) => {
    setSelectedValue(value)
    setIsOpen(false)
  }

  useEffect(() => {
    onChange(selectedValue)
  }, [selectedValue])

  useEffect(() => {
    function handleIsClickOutside(event) {
      if (
        isOpen &&
        dropdownMenu.current &&
        !dropdownMenu.current.contains(event.target)
      ) {
        setIsOpen(false)
      }
    }
    document.addEventListener('mousedown', handleIsClickOutside, true)
    return () => {
      document.removeEventListener('mousedown', handleIsClickOutside, true)
    }
  }, [isOpen])

  return (
    <div className="flex flex-row items-center">
      <div className="mr-2 flex items-center justify-center">{label}</div>
      <div className="block" ref={dropdownMenu}>
        <button
          type="button"
          className="flex h-6 items-center justify-between rounded-md border border-slate-100 pl-2 text-base font-thin focus:border-slate-300 focus:outline-none"
          onClick={() => setIsOpen((oldState) => !oldState)}
        >
          {selectedValue}
          {openClosedIcon(isOpen)}
        </button>
        {isOpen && (
          <DropdownItemsList
            className="mt-1 max-h-28 w-28 cursor-pointer border border-gray-200"
            items={items}
            selectedValue={selectedValue}
            handleSelect={handleSelectItem}
          />
        )}
      </div>
    </div>
  )
}

export default Dropdown
