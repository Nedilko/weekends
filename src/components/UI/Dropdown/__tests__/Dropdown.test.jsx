import Dropdown from '@UI/Dropdown/Dropdown'
import { render, screen, vi, userEvent } from '@utils/test-utils'

describe('dropdown', () => {
  const items = ['first', 'second', 'third']
  beforeAll(() => {
    window.HTMLElement.prototype.scrollIntoView = vi.fn()
  })

  afterEach(() => {
    window.HTMLElement.prototype.scrollIntoView.mockClear()
  })

  it('should render closed', () => {
    const handleChage = vi.fn()
    const value = 'second'
    render(
      <Dropdown
        label="sample label"
        items={items}
        onChange={handleChage}
        value={value}
      />
    )
    const dropdown = screen.getByText('sample label')
    expect(dropdown).toMatchSnapshot()
  })

  it('should render opened', async () => {
    const handleChage = vi.fn()
    const value = 'second'
    const user = userEvent.setup()
    const { container } = render(
      <Dropdown
        label="sample label"
        items={items}
        onChange={handleChage}
        value={value}
      />
    )
    const dropdownElement = screen.getByRole('button')
    await user.click(dropdownElement)
    expect(container).toMatchSnapshot()
  })

  it('should render closed on click outside', async () => {
    const handleChage = vi.fn()
    const value = 'second'
    const user = userEvent.setup()
    const { container } = render(
      <Dropdown
        label="sample label"
        items={items}
        onChange={handleChage}
        value={value}
      />
    )
    await user.click(document.body)
    expect(container).toMatchSnapshot()
  })

  it('should change selected value', async () => {
    const handleChage = vi.fn()
    const value = 'second'
    const user = userEvent.setup()
    render(
      <Dropdown
        label="sample label"
        items={items}
        onChange={handleChage}
        value={value}
      />
    )
    const dropdownElement = screen.getByRole('button')
    await user.click(dropdownElement)
    const dropdownItem = screen.getByText('third')
    await user.click(dropdownItem)
    expect(dropdownElement).toHaveTextContent('third')
  })

  it('should close dropdown on value select', async () => {
    const handleChage = vi.fn()
    const value = 'second'
    const user = userEvent.setup()
    render(
      <Dropdown
        label="sample label"
        items={items}
        onChange={handleChage}
        value={value}
      />
    )
    await user.click(screen.getByRole('button'))
    const dropdownItemsList = screen.queryAllByTestId('dropdown-item')
    await user.click(screen.getByText('third'))
    dropdownItemsList.forEach((item) => expect(item).not.toBeInTheDocument())
  })

  it('should rotate arrow when opened', async () => {
    const handleChage = vi.fn()
    const value = 'second'
    const user = userEvent.setup()
    render(
      <Dropdown
        label="sample label"
        items={items}
        onChange={handleChage}
        value={value}
      />
    )
    const arrow = screen.getByRole('img')
    expect(arrow).toMatchSnapshot()
    await user.click(screen.getByRole('button'))
    expect(arrow).toMatchSnapshot()
  })
})
