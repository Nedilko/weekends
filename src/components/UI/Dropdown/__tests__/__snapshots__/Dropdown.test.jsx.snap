// Vitest Snapshot v1

exports[`Dropdown > should render closed 1`] = `
<div
  class="mr-2 flex items-center justify-center"
>
  sample label
</div>
`;

exports[`Dropdown > should render closed on click outside 1`] = `
<div>
  <div
    class="flex flex-row items-center"
  >
    <div
      class="mr-2 flex items-center justify-center"
    >
      sample label
    </div>
    <div
      class="z-10 block"
    >
      <button
        class="flex h-6 items-center justify-between rounded-md border border-slate-100 pl-2 text-base font-thin focus:border-slate-300 focus:outline-none dark:border-gray-700 dark:focus:border-gray-500"
        type="button"
      >
        second
        <svg
          class="mr-1 h-4 w-4 text-gray-500 "
          fill="none"
          role="img"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Dropdown > should render opened 1`] = `
<div>
  <div
    class="flex flex-row items-center"
  >
    <div
      class="mr-2 flex items-center justify-center"
    >
      sample label
    </div>
    <div
      class="z-10 block"
    >
      <button
        class="flex h-6 items-center justify-between rounded-md border border-slate-100 pl-2 text-base font-thin focus:border-slate-300 focus:outline-none dark:border-gray-700 dark:focus:border-gray-500"
        type="button"
      >
        second
        <svg
          class="mr-1 h-4 w-4 text-gray-500 rotate-180"
          fill="none"
          role="img"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <div
        class="absolute overflow-auto rounded-md bg-white shadow-md dark:bg-zinc-900 mt-1 max-h-28 w-28 cursor-pointer border border-gray-200 dark:border-gray-500"
      >
        <div
          class="px-2 py-1 text-sm hover:bg-slate-200 dark:hover:bg-slate-700 font-thin"
          data-testid="dropdown-item"
          value="first"
        >
          first
        </div>
        <div
          class="px-2 py-1 text-sm hover:bg-slate-200 dark:hover:bg-slate-700 font-regular"
          data-testid="dropdown-item"
          value="second"
        >
          second
        </div>
        <div
          class="px-2 py-1 text-sm hover:bg-slate-200 dark:hover:bg-slate-700 font-thin"
          data-testid="dropdown-item"
          value="third"
        >
          third
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown > should rotate arrow when opened 1`] = `
<svg
  class="mr-1 h-4 w-4 text-gray-500 "
  fill="none"
  role="img"
  stroke="currentColor"
  stroke-width="2"
  viewBox="0 0 24 24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M19 9l-7 7-7-7"
    stroke-linecap="round"
    stroke-linejoin="round"
  />
</svg>
`;

exports[`Dropdown > should rotate arrow when opened 2`] = `
<svg
  class="mr-1 h-4 w-4 text-gray-500 rotate-180"
  fill="none"
  role="img"
  stroke="currentColor"
  stroke-width="2"
  viewBox="0 0 24 24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M19 9l-7 7-7-7"
    stroke-linecap="round"
    stroke-linejoin="round"
  />
</svg>
`;
